### PR TITLE
Feat/disclose errors

### PR DESCRIPTION
**Pull request by RomanTeren**

### CODE REVIEW

File: src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java [hunk 1]
```diff
@@ -108,24 +114,49 @@ private OneTimePassword getOneTimePassword(OneTimePasswordDto oneTimePasswordDto
...
+            .findById(oneTimePasswordCheckDto.getId())
             .orElseThrow(OtpInvalidPasswordException::new);
-
-        if (checkOtpState(otp)
-            || checkOtpDate(otp)
-            || checkOtpRetries(otp)
-            || checkOtpPasswd(otp, oneTimePasswordCheckDto)) {
-
+        try {
+            if (checkOtpState(otp)) {
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java:122]: please throw each exception inside each of these `check` methods

example: 
```java 
    private void checkOtpState(OneTimePassword otp) {
        if (otp.getStateKey() != StateKey.ACTIVE) {
              throw new IllegalOtpStateException();
        } 
    }
```

it'll result in much cleaner code here in main `check` method

File: src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java [hunk 2]
```diff
@@ -108,24 +114,49 @@ private OneTimePassword getOneTimePassword(OneTimePasswordDto oneTimePasswordDto
...
+            Boolean discloseCheckErrorsEnabled = ofNullable(otp)
+                .map(OneTimePassword::getTypeKey)
+                .map(otpSpecService::getOtpTypeSpec)
+                .map(OtpTypeSpec::getDiscloseCheckErrors)
+                .orElse(false);
+
+            if (discloseCheckErrorsEnabled) {
+                throw exception;
+            } else {
+                throw new InvalidPasswordException();
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java:154]: I suggest adding some log message for investigating purposes

File: src/main/java/com/icthh/xm/ms/otp/web/rest/errors/ExpiredOtpException.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.icthh.xm.ms.otp.web.rest.errors;
+
+import static com.icthh.xm.ms.otp.web.rest.errors.ErrorConstants.EXPIRED_OTP;
+
+public class ExpiredOtpException extends InvalidPasswordException {
+
+    public ExpiredOtpException() {
+        super("OTP expired", EXPIRED_OTP);
+    }
+}
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/otp/web/rest/errors/ExpiredOtpException.java:10]: please add newline

File: src/main/java/com/icthh/xm/ms/otp/web/rest/errors/OtpPasswordNotMatchException.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.icthh.xm.ms.otp.web.rest.errors;
+
+import static com.icthh.xm.ms.otp.web.rest.errors.ErrorConstants.OTP_PASSWORD_NOT_MATCH;
+
+public class OtpPasswordNotMatchException extends InvalidPasswordException {
+
+    public OtpPasswordNotMatchException() {
+        super("OTP Password is not correct", OTP_PASSWORD_NOT_MATCH);
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/otp/web/rest/errors/OtpPasswordNotMatchException.java:8]: Tautology detected 

please update to `OTP is not correct`

File: src/main/java/com/icthh/xm/ms/otp/web/rest/errors/OtpPasswordNotMatchException.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.icthh.xm.ms.otp.web.rest.errors;
+
+import static com.icthh.xm.ms.otp.web.rest.errors.ErrorConstants.OTP_PASSWORD_NOT_MATCH;
+
+public class OtpPasswordNotMatchException extends InvalidPasswordException {
```
Comment (RomanTeren) [inline src/main/java/com/icthh/xm/ms/otp/web/rest/errors/OtpPasswordNotMatchException.java:5]: OtpNotMatchedException

File: src/main/java/com/icthh/xm/ms/otp/domain/OtpSpec.java [hunk 6]
```diff
@@ -71,6 +85,6 @@
         /**
          * Invocation limit value
          */
-        private Integer value;
+        private Integer value; // 5
```
[the remark this answers is not in the document]
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/otp/domain/OtpSpec.java]: fixed

File: src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java [hunk 7]
```diff
@@ -108,24 +114,49 @@ private OneTimePassword getOneTimePassword(OneTimePasswordDto oneTimePasswordDto
 
     @Override
     public void check(OneTimePasswordCheckDto oneTimePasswordCheckDto) {
-
-        OneTimePassword otp = oneTimePasswordRepository.findById(oneTimePasswordCheckDto.getId())
+        // validate
```
[the remark this answers is not in the document]
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java]: fixed

File: src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java (hunk 1 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/otp/service/impl/OneTimePasswordServiceImpl.java]: fixed

File: src/main/java/com/icthh/xm/ms/otp/web/rest/errors/ExpiredOtpException.java (hunk 3 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/otp/web/rest/errors/ExpiredOtpException.java]: fixed

File: src/main/java/com/icthh/xm/ms/otp/web/rest/errors/OtpPasswordNotMatchException.java (hunk 5 above)
Comment (illiaredko) [reply src/main/java/com/icthh/xm/ms/otp/web/rest/errors/OtpPasswordNotMatchException.java]: fixed

**Review by major13ua — APPROVED** [review]

**Review by sergeysenja1992 — APPROVED** [review]

**Pull request merged**